### ISSUE TITLE
Is it possible to have a Merge function with two mode mul and sum working sequentially?

### ISSUE THREAD

**Issue by mmoradik** [fchollet/keras#3225]
I am new in keras. I have some problem working with merge.

Assuming that I have 3 models. model1 has 4 outputs (px1,py1,px2,py2).  model 2 and models 3 each  has two outputs. (x1,y1) and (x2,y2) are outputs of each one of them respectively.
My question is how I can merge these models to have two outputs x,y:
x=px1*x1+px2*x2
y=py1*y1+py2*y2

Through the following code, the output of final model has 4 dimension representing (px1*x1, px2*x2, py1*y1, py2*y2). How can I reshape this output and sum to reach to x and y.
or is there any way to merge only some outputs of a model with other model?

model1=Sequential()
model1.add(Dense(input_dim=10, output_dim=2*2))
model1.add(Activation('softmax'))

model2=Sequential()
model2.add(Dense(input_dim=10, output_dim=2))
model2.add(Activation('tanh'))

model3=Sequential()
model3.add(Dense(input_dim=10, output_dim=2))
model3.add(Activation('tanh'))

merged_expert=Merge([model2,model3],mode='concat')
merged = Merge([model1, merged_expert], mode='mul')

final_model = Sequential()
final_model.add(merged)

**Comment by farizrahman4u** [comment]
In complex cases such as this one, its better to use a custom merge function than multiple standard Merges like sum and concat. This will let you change your math without changing your model structure.

```python

from keras import backend as K


def f(M):
    m1, m2, m3 = M
    z = m1 * K.concatenate([m2, m3])
    x = K.sum(z[:, :2], axis=1)
    y = K.sum(z[:, 2:], axis=1)
    return K.concatenate([x, y])

model1=Sequential()
model1.add(Dense(input_dim=10, output_dim=2*2))
model1.add(Activation('softmax'))

model2=Sequential()
model2.add(Dense(input_dim=10, output_dim=2))
model2.add(Activation('tanh'))

model3=Sequential()
model3.add(Dense(input_dim=10, output_dim=2))
model3.add(Activation('tanh'))

merged = Merge([model1, model2, model3], mode=f, output_shape=(2,))

final_model = Sequential()
final_model.add(merged)

```

**Comment by mmoradik** [comment]
Thank you for you response. I receive an error mentioning "Invalid merge mode: <function f at 0x7f4945cd0320>"

if mode not in {'sum', 'mul', 'concat', 'ave', 'join', 'cos', 'dot'}:   raise Exception('Invalid merge mode: ' + str(mode))

**Comment by mmoradik** [comment]
by updating the keras, model is successfully compiled. But during training, another error appeared:
operands could not be broadcast together with shapes (1,64) (32,2) 
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.

**Comment by mmoradik** [comment]
The shape of the output in this case is:
model.output_shape=((None, 4), 2)

while it supposed to be (None, 2) !!!

**Comment by farizrahman4u** [comment]
I have updated the code. Try again.

**Comment by mmoradik** [comment]
Thank you so much, now it works.

**Comment by farizrahman4u** [comment]
Cool, close the issue if everythings fine.

**Comment by mmoradik** [comment]
What about a case where model1 has 2 outputs (p1,p2) and model2 and models3 each has two outputs (x1,y1) and (x2,y2) respectively.
How we can change the code to get:
 x=p1x1+p2x2
 y=p1y1+p2y2
Basically, fist output of model 1 need to be multiplied in all outputs of model2.
and the second output of model1 need to be multiplied in all outputs of model3.

Thank you

**Comment by farizrahman4u** [comment]
Just write the corresponding function. 

```python
def f(M):
    m1, m2, m3 = M
    # Optimization : Do all the 4 multiplications at once, do both the additions at once
    z = K.concatenate([m1] * 2) * K.concatenate([m2[:, :1], m3[:, :1], m2[:, 1:], m3[:, 1:]])
    z = K.reshape(z, (-1, 2, 2))
    return K.sum(z, 1)
```

P.S Please try out pure math stuff yourself. I will be happy to solve any issue you might have with Keras. Hope you understand.

**Comment by alyato** [comment]
@farizrahman4u  Thanks for  your suggestion. There are some question i can't understand.

> def f(M):
    m1, m2, m3 = M
    Y = m1 * K.concatenate([m2, m3])
    Y = K.reshape(Y, (-1, 2, 2))
    Y = K.sum(Y, 2)
    return Y

The model_1 is (none ,4),model_2 is (none,2),model_3 is (none , 2)

>   K.concatenate([m2, m3]),after that ,we can get the output_shape( none,4)
 Then  Y = m1 * K.concatenate([m2, m3]) , we can get the output_shape(none,4*4)
Then  Y = K.reshape(Y, (-1, 2, 2)), the Y_shape is (none,4,2,2)

But i don't know what donest the K.sum do?
I also check the source code and don't understand.
![sum](https://cloud.githubusercontent.com/assets/6815071/17206258/1870c2b6-546c-11e6-9d7c-ae36bc354662.png)
The Y_shape is (none,4,2,2), and what don't the K.sum(Y,2) do?
Thanks.

**Comment by farizrahman4u** [comment]
You got it all wrong, * is an element wise operation, both its operators should have the same shape and the resultant would have the same shape as the inputs.

**Comment by farizrahman4u** [comment]
So Y's shape would be (None, 4).. And when you do K.reshape, the rank of resultant tensor would be the lengh of the shape specified (obviously),  there is no way a 4 dim tensor would be created anywhere in the above code.

**Comment by alyato** [comment]
@farizrahman4u  Thanks.  i got it. The Y's shape is (none ,4) after `Y = m1 * K.concatenate([m2, m3])`
I don't understand the code `Y = K.reshape(Y, (-1, 2, 2))`.
Does the shape of Y is (none,1,2,2)? because the Y's shape is (none ,4) and it totally has 4. if calling the 
function of Reshape,the Y became  (none,1,2,2) and it totally also has 4.
Does it right? Thanks.

**Comment by alyato** [comment]
@farizrahman4u  Is the numpy.reshape different from the K.reshape?
I check the numpy.reshape and it give one example.
![res](https://cloud.githubusercontent.com/assets/6815071/17234277/9080fc5c-54f4-11e6-9a38-0b1a411663ea.png)
And the K.reshape 
![re](https://cloud.githubusercontent.com/assets/6815071/17234296/bfd2a1e0-54f4-11e6-9fce-c43ef085a91c.png)
So what does your mean?

> there is no way a 4 dim tensor would be created anywhere in the above code.

**Comment by farizrahman4u** [comment]
Y's shape was (None, 4). After reshaping it to (-1, 2, 2) it becomes (None, 2, 2). Its that simple. Not sure what's confusing you.

**Comment by alyato** [comment]
@farizrahman4u Thanks for your patient. What confused me is the value `-1` in (-1,2,2). I don't know why set the `-1`. when Y call the `K.reshape(Y, (-1, 2, 2))`,  i think that the Y become (none ,1,2,2). Besides i don't know why axis = 2 in the `Y = K.sum(Y, 2)`. Does the axis be set 3 or 4? Thanks

**Comment by farizrahman4u** [comment]
-1 is for unknown dimension. axis = 2 because I am summing over the time dimension.

**Comment by dimahdera** [comment]
What does K.concatenate really do?
